src/Data/Morpheus/Types/Internal/Subscription/ClientStore.hs

```diff
@@ -0,0 +1,132 @@
+
+module Data.Morpheus.Types.Internal.Subscription.ClientStore
+  ( Client(..)
+  , ID
+  , ClientStore
+  , SesionID
+  , empty
+  , insert
+  , adjust
+  , delete
+  , publish
+  )
+where
+
+import           Data.List                      ( intersect )
+import           Data.Foldable                  ( traverse_ )
+import           Data.ByteString.Lazy.Char8     (ByteString)
+import           Data.Semigroup                 ( (<>) )
+import           Data.Text                      ( Text )
+import           Data.UUID                      ( UUID )
+import           Data.HashMap.Lazy              ( HashMap , keys)
+import qualified Data.HashMap.Lazy   as  HM     ( empty
+                                                , insert
+                                                , delete
+                                                , adjust
+                                                , elems
+                                                , toList
+                                                )
+
+
+-- MORPHEUS
+import           Data.Morpheus.Types.Internal.Subscription.Apollo
+                                                ( toApolloResponse
+                                                )
+import           Data.Morpheus.Types.Internal.Operation
+                                                (Empty(..))
+import           Data.Morpheus.Types.Internal.Resolving
+                                                ( SubEvent 
+                                                , Event(..)
+                                                , GQLChannel(..)
+                                                )
+
+type ID = UUID
+
+type SesionID = Text
+
+data Client e ( m :: * -> * ) =
+  Client
+    { clientId       :: ID
+    , clientCallback :: ByteString -> m ()
+    , clientSessions :: HashMap SesionID (SubEvent e m)
+    }
+
+instance Show (Client e m) where
+  show Client { clientId, clientSessions } =
+    "Client { id: "
+      <> show clientId
+      <> ", sessions: "
+      <> show (keys clientSessions)
+      <> " }"
+
+publish
+  :: ( Eq (StreamChannel event)
+     , GQLChannel event
+     , Monad m
+     ) 
+  => event 
+  -> ClientStore event m 
+  -> m ()
+publish event = traverse_ sendMessage . elems
+ where
+  sendMessage Client { clientSessions, clientCallback }
+    | null clientSessions  = pure ()
+    | otherwise = traverse_ send (filterByChannels clientSessions)
+   where
+    send (sid, Event { content = subscriptionRes }) 
+      = toApolloResponse sid <$> subscriptionRes event >>= clientCallback
+    ---------------------------
+    filterByChannels = filter
+      ( not
+      . null
+      . intersect (streamChannels event)
+      . channels
+      . snd
+      ) . HM.toList
+
+-- subscription 
+-- store
+-- stores active subsciption connections and requests
+newtype ClientStore e ( m :: * -> * ) = 
+    ClientStore 
+      { unpackStore :: HashMap ID (Client e m)
+      } deriving (Show)
```

Comment:
This one is still a little hard to understand for me. It seems like a mapping between websocket sessionIds to something ? But I don't see why it's necessary, what is its purpose ?

Reply:
it is not sessionID. it is client ID to remove client after the Connection ended.

Reply:
But why do you need to track the clients server side ? Why is the open client-server connection not enough ?

Reply:
Oh ok I understand now, yes ClientConnectionStore is probably better. So the server has to track the current open connections with all its clients and that's the purpose of ClientConnectionStore.